File: frontend/lib/model/lichlamviec/lich_lam_viec.dart
class LichLamViec {
  final int? maLich;
  final int maNV;
  final String ngayLamViec; // yyyy-MM-dd
  final String? gioBatDau; // HH:mm:ss
  final String? gioKetThuc; // HH:mm:ss
  final String? caLamViec; // Sáng, Chiều, Tối
  final String? loaiCa; // Bình thường, Tăng ca, Làm thêm
  final String? ghiChu;
  final String trangThai; // KICH_HOAT, HUY
  final String? ngayTao; // yyyy-MM-dd HH:mm:ss
  final String? ngayCapNhat; // yyyy-MM-dd HH:mm:ss
  final String? nguoiTao;
  final bool daXoa;

  LichLamViec({
    this.maLich,
    required this.maNV,
    required this.ngayLamViec,
    this.gioBatDau,
    this.gioKetThuc,
    this.caLamViec,
    this.loaiCa,
    this.ghiChu,
    this.trangThai = 'KICH_HOAT',
    this.ngayTao,
    this.ngayCapNhat,
    this.nguoiTao,
    this.daXoa = false,
  });

  factory LichLamViec.fromJson(Map<String, dynamic> json) {
    return LichLamViec(
      maLich: json['maLich'],
      maNV: json['maNV'],
      ngayLamViec: json['ngayLamViec'],
      gioBatDau: json['gioBatDau'],
      gioKetThuc: json['gioKetThuc'],
      caLamViec: json['caLamViec'],
      loaiCa: json['loaiCa'],
      ghiChu: json['ghiChu'],
      trangThai: json['trangThai'] ?? 'KICH_HOAT',
      ngayTao: json['ngayTao'],
      ngayCapNhat: json['ngayCapNhat'],
      nguoiTao: json['nguoiTao'],
      daXoa: json['daXoa'] ?? false,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'maLich': maLich,
      'maNV': maNV,
      'ngayLamViec': ngayLamViec,
      'gioBatDau': gioBatDau,
      'gioKetThuc': gioKetThuc,
      'caLamViec': caLamViec,
      'loaiCa': loaiCa,
      'ghiChu': ghiChu,
      'trangThai': trangThai,
      'ngayTao': ngayTao,
      'ngayCapNhat': ngayCapNhat,
      'nguoiTao': nguoiTao,
      'daXoa': daXoa,
    };
  }

  // Helper methods
  DateTime get ngayLamViecAsDateTime => DateTime.parse(ngayLamViec);

  DateTime? get gioBatDauAsTime {
    if (gioBatDau == null) return null;
    final parts = gioBatDau!.split(':');
    return DateTime(2000, 1, 1, int.parse(parts[0]), int.parse(parts[1]), int.parse(parts[2]));
  }

  DateTime? get gioKetThucAsTime {
    if (gioKetThuc == null) return null;
    final parts = gioKetThuc!.split(':');
    return DateTime(2000, 1, 1, int.parse(parts[0]), int.parse(parts[1]), int.parse(parts[2]));
  }

  String get caLamViecDisplay {
    switch (caLamViec) {
      case 'SANG':
        return 'Ca Sáng';
      case 'CHIEU':
        return 'Ca Chiều';
      case 'TOI':
        return 'Ca Tối';
      default:
        return caLamViec ?? '';
    }
  }

  String get loaiCaDisplay {
    switch (loaiCa) {
      case 'BINH_THUONG':
        return 'Bình thường';
      case 'TANG_CA':
        return 'Tăng ca';
      case 'LAM_THEM':
        return 'Làm thêm';
      default:
        return loaiCa ?? '';
    }
  }

  String get trangThaiDisplay {
    switch (trangThai) {
      case 'KICH_HOAT':
        return 'Kích hoạt';
      case 'HUY':
        return 'Hủy';
      default:
        return trangThai;
    }
  }

  LichLamViec copyWith({
    int? maLich,
    int? maNV,
    String? ngayLamViec,
    String? gioBatDau,
    String? gioKetThuc,
    String? caLamViec,
    String? loaiCa,
    String? ghiChu,
    String? trangThai,
    String? ngayTao,
    String? ngayCapNhat,
    String? nguoiTao,
    bool? daXoa,
  }) {
    return LichLamViec(
      maLich: maLich ?? this.maLich,
      maNV: maNV ?? this.maNV,
      ngayLamViec: ngayLamViec ?? this.ngayLamViec,
      gioBatDau: gioBatDau ?? this.gioBatDau,
      gioKetThuc: gioKetThuc ?? this.gioKetThuc,
      caLamViec: caLamViec ?? this.caLamViec,
      loaiCa: loaiCa ?? this.loaiCa,
      ghiChu: ghiChu ?? this.ghiChu,
      trangThai: trangThai ?? this.trangThai,
      ngayTao: ngayTao ?? this.ngayTao,
      ngayCapNhat: ngayCapNhat ?? this.ngayCapNhat,
      nguoiTao: nguoiTao ?? this.nguoiTao,
      daXoa: daXoa ?? this.daXoa,
    );
  }
}

File: backend/src/main/java/com/example/worktrack/model/lichlamviec/LichLamViec.java
package com.example.worktrack.model.lichlamviec;

import com.fasterxml.jackson.annotation.JsonFormat;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;

@Entity
@Table(name = "lich_lam_viec")
@Data
@NoArgsConstructor
@AllArgsConstructor
public class LichLamViec {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "ma_lich")
    private Integer maLich;

    @Column(name = "ma_nv", nullable = false)
    private Integer maNV;

    @Column(name = "ngay_lam_viec", nullable = false)
    @JsonFormat(pattern = "yyyy-MM-dd")
    private LocalDate ngayLamViec;

    @Column(name = "gio_bat_dau")
    @JsonFormat(pattern = "HH:mm:ss")
    private LocalTime gioBatDau;

    @Column(name = "gio_ket_thuc")
    @JsonFormat(pattern = "HH:mm:ss")
    private LocalTime gioKetThuc;

    @Column(name = "ca_lam_viec", length = 50)
    private String caLamViec; // Sáng, Chiều, Tối

    @Column(name = "loai_ca", length = 50)
    private String loaiCa; // Bình thường, Tăng ca, Làm thêm

    @Column(name = "ghi_chu", columnDefinition = "TEXT")
    private String ghiChu;

    @Column(name = "trang_thai", length = 20, nullable = false)
    private String trangThai = "KICH_HOAT"; // KICH_HOAT, HUY

    @Column(name = "ngay_tao", nullable = false)
    @JsonFormat(pattern = "yyyy-MM-dd HH:mm:ss")
    private LocalDateTime ngayTao;

    @Column(name = "ngay_cap_nhat")
    @JsonFormat(pattern = "yyyy-MM-dd HH:mm:ss")
    private LocalDateTime ngayCapNhat;

    @Column(name = "nguoi_tao")
    private String nguoiTao;

    @Column(name = "da_xoa", nullable = false)
    private Boolean daXoa = false;

    @PrePersist
    protected void onCreate() {
        ngayTao = LocalDateTime.now();
        ngayCapNhat = LocalDateTime.now();
    }

    @PreUpdate
    protected void onUpdate() {
        ngayCapNhat = LocalDateTime.now();
    }
}